File: backend/src/controllers/authController.ts
import type { Request, Response } from 'express';
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { getDB } from '../db/database.js';

const JWT_SECRET = process.env.JWT_SECRET || 'diquan_secret_key';

export const register = async (req: Request, res: Response) => {
    const { phone, password, nickname, role } = req.body;
    const db = getDB();

    try {
        if (!phone || !password) {
            return res.status(400).json({ error: '手机号和密码不能为空' });
        }
        const hashedPassword = await bcrypt.hash(password, 10);
        const safeRole = role === 'MERCHANT' ? 'MERCHANT' : 'USER';
        await db.run(
            'INSERT INTO users (phone, password, nickname, role) VALUES (?, ?, ?, ?)',
            [phone, hashedPassword, nickname || phone, safeRole]
        );
        res.status(201).json({ message: '注册成功' });
    } catch (error: any) {
        if (error.message.includes('UNIQUE constraint failed')) {
            res.status(400).json({ error: '手机号已注册' });
        } else {
            res.status(500).json({ error: '注册失败' });
        }
    }
};

export const login = async (req: Request, res: Response) => {
    const { phone, password } = req.body;
    const db = getDB();

    try {
        const user = await db.get('SELECT * FROM users WHERE phone = ?', [phone]);
        if (!user || user.status === 'DISABLED' || !(await bcrypt.compare(password, user.password))) {
            return res.status(401).json({ error: '手机号或密码错误' });
        }

        const token = jwt.sign({ id: user.id, role: user.role, phone: user.phone }, JWT_SECRET, { expiresIn: '7d' });
        await db.run('INSERT INTO audit_logs (user_id, action, target_type, target_id, detail) VALUES (?, ?, ?, ?, ?)', [user.id, 'LOGIN', 'users', user.id, '用户登录']);
        res.json({
            token,
            user: {
                id: user.id,
                phone: user.phone,
                nickname: user.nickname,
                role: user.role,
                avatar: user.avatar,
                enterprise_status: user.enterprise_status
            }
        });
    } catch (error) {
        res.status(500).json({ error: '登录失败' });
    }
};

File: backend/src/main/java/com/diquan/backend/AuthController.java
package com.diquan.backend;

import org.springframework.dao.DuplicateKeyException;
import org.springframework.http.HttpStatus;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RestController;

import java.util.LinkedHashMap;
import java.util.Map;

@RestController
@RequestMapping("/api/auth")
class AuthController {
    private final Sql sql;
    private final BCryptPasswordEncoder passwordEncoder;
    private final AuthService authService;

    AuthController(Sql sql, BCryptPasswordEncoder passwordEncoder, AuthService authService) {
        this.sql = sql;
        this.passwordEncoder = passwordEncoder;
        this.authService = authService;
    }

    @PostMapping("/register")
    Map<String, String> register(@RequestBody Map<String, Object> body) {
        String phone = Values.text(body, "phone");
        String password = Values.text(body, "password");
        if (!Values.hasText(phone) || !Values.hasText(password)) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "手机号和密码不能为空");
        }

        String role = "MERCHANT".equals(Values.text(body, "role")) ? "MERCHANT" : "USER";
        try {
            sql.update(
                "INSERT INTO users (phone, password, nickname, role) VALUES (?, ?, ?, ?)",
                phone,
                passwordEncoder.encode(password),
                Values.text(body, "nickname", phone),
                role
            );
            return Map.of("message", "注册成功");
        } catch (DuplicateKeyException error) {
            throw new ApiException(HttpStatus.BAD_REQUEST, "手机号已注册");
        }
    }

    @PostMapping("/login")
    Map<String, Object> login(@RequestBody Map<String, Object> body) {
        String phone = Values.text(body, "phone");
        String password = Values.text(body, "password");
        Map<String, Object> user = sql.one("SELECT * FROM users WHERE phone = ?", phone);
        if (user == null
            || "DISABLED".equals(user.get("status"))
            || !passwordEncoder.matches(password == null ? "" : password, String.valueOf(user.get("password")))) {
            throw new ApiException(HttpStatus.UNAUTHORIZED, "手机号或密码错误");
        }

        long userId = Values.longValue(user.get("id"));
        String role = String.valueOf(user.get("role"));
        sql.update(
            "INSERT INTO audit_logs (user_id, action, target_type, target_id, detail) VALUES (?, ?, ?, ?, ?)",
            userId,
            "LOGIN",
            "users",
            userId,
            "用户登录"
        );

        Map<String, Object> safeUser = new LinkedHashMap<>();
        safeUser.put("id", userId);
        safeUser.put("phone", user.get("phone"));
        safeUser.put("nickname", user.get("nickname"));
        safeUser.put("role", role);
        safeUser.put("avatar", user.get("avatar"));
        safeUser.put("enterprise_status", user.get("enterprise_status"));

        Map<String, Object> response = new LinkedHashMap<>();
        response.put("token", authService.token(userId, role, String.valueOf(user.get("phone"))));
        response.put("user", safeUser);
        return response;
    }
}
